### PR TITLE
feat: add `spancheck` linter

### LINKED ISSUE #3722: Add "opentelemetry" linter
### Your feature request related to a problem? Please describe.

The `go-opentelemetry-lint` linter helps with properly using the OpenTelemetry SDK for Go.

```go
// bad
func Query(ctx context.Context, db *sql.DB) error {
	row := db.QueryRowContext(ctx, `SELECT * FROM sample_text`)
	return row.Err()
}

// good
func Query(ctx context.Context, db *sql.DB) error {
	ctx, span := tracer().Start(ctx, "query")
	defer span.End()

	row := db.QueryRowContext(ctx, `SELECT * FROM sample_text`)
	return row.Err()
}
```

https://github.com/polyfloyd/go-opentelemetry-lint

### Describe the solution you'd like.

—

### Describe alternatives you've considered.

—

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by jjti**
## Background

This adds a `spancheck` linter: https://github.com/jjti/go-spancheck

The linter checks usage of OpenTelemetry spans, which are easy to get wrong. Self-instrumented traces requires a lot of easy-to-forget boilerplate:

```go
import (
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/codes"
)

func task(ctx context.Context) error {
    ctx, span := otel.Tracer("foo").Start(ctx, "bar")
    defer span.End() // call `.End()`

    if err := subTask(ctx); err != nil {
        span.SetStatus(codes.Error, err.Error()) // call SetStatus(codes.Error, msg) to set status:error
        span.RecordError(err) // call RecordError(err) to record an error event
        return err
    }

    return nil
}
```

Checks for spans include:

1. checking that `span.End()` is called (by default, not having this can cause memory leaks, and spans never being closed/useful)
1. optionally checking `span.SetStatus(codes.Error, msg)` is called on error
1. optionally checking `span.RecordError(err)` is called on error

## Testing

[Like `protogetter`](https://github.com/golangci/golangci-lint/pull/4069/files) I have an external dependency (otel) so made `spanchecker`s tests their own module in `test/testdata/spancheck/spancheck.go`

```bash
$ T=spancheck make test_linters 
GL_TEST_RUN=1 go test -v ./test -count 1 -run TestSourcesFromTestdata/spancheck
=== RUN   TestSourcesFromTestdata
    linters_test.go:21: testdata/*.go
--- PASS: TestSourcesFromTestdata (0.30s)
=== RUN   TestSourcesFromTestdataSubDir
=== RUN   TestSourcesFromTestdataSubDir/spancheck
    linters_test.go:41: testdata/spancheck/*.go
=== RUN   TestSourcesFromTestdataSubDir/spancheck/spancheck_default.go
=== PAUSE TestSourcesFromTestdataSubDir/spancheck/spancheck_default.go
=== RUN   TestSourcesFromTestdataSubDir/spancheck/spancheck_enable_all_copy.go
=== PAUSE TestSourcesFromTestdataSubDir/spancheck/spancheck_enable_all_copy.go
=== CONT  TestSourcesFromTestdataSubDir/spancheck/spancheck_default.go
=== CONT  TestSourcesFromTestdataSubDir/spancheck/spancheck_enable_all_copy.go
level=info msg="[test] ran [/Users/josh/GitHub/golangci/golangci-lint/golangci-lint run --internal-cmd-test -c configs/enable_all.yml --allow-parallel-runners --disable-all --out-format=json --max-same-issues=100 --disable-all -Espancheck spancheck_enable_all copy.go] in 734.076417ms"
level=info msg="[test] ran [/Users/josh/GitHub/golangci/golangci-lint/golangci-lint run --internal-cmd-test -c configs/default.yml --allow-parallel-runners --disable-all --out-format=json --max-same-issues=100 --disable-all -Espancheck spancheck_default.go] in 770.173792ms"
level=info msg="[test] ran [/Users/josh/GitHub/golangci/golangci-lint/golangci-lint run --internal-cmd-test -c configs/enable_all.yml --allow-parallel-runners --disable-all --out-format=json --max-same-issues=100 -Etypecheck --disable-all -Espancheck spancheck_enable_all copy.go] in 307.45575ms"
level=info msg="[test] ran [/Users/josh/GitHub/golangci/golangci-lint/golangci-lint run --internal-cmd-test -c configs/default.yml --allow-parallel-runners --disable-all --out-format=json --max-same-issues=100 -Etypecheck --disable-all -Espancheck spancheck_default.go] in 322.52025ms"
--- PASS: TestSourcesFromTestdataSubDir (1.40s)
    --- PASS: TestSourcesFromTestdataSubDir/spancheck (0.31s)
        --- PASS: TestSourcesFromTestdataSubDir/spancheck/spancheck_enable_all_copy.go (1.04s)
        --- PASS: TestSourcesFromTestdataSubDir/spancheck/spancheck_default.go (1.09s)
PASS
ok      github.com/golangci/golangci-lint/test  2.112s
```

Fixes #3722

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/golangci/golangci-lint
 
-go 1.20
+go 1.21.3
```
Comment (jjti) [inline go.mod:3]: note to self: wat

**Comment by ldez** [comment]
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

- [ ] The CLA must be signed

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description of the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter. (the team will help to verify that)
- [ ] It must have a valid license (AGPL is not allowed) and the file must contain the required information by the license, ex: author, year, etc.
- [ ] The linter repository must have a CI and tests.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`.
- [ ] It must not contain `log.fatal()`, `os.exit()`, or similar.
- [ ] It must not modify the AST.
- [ ] It must not have false positives/negatives. (the team will help to verify that)
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must work with `T=<name of the linter test file>.go make test_linters`. (the team will help to verify that)

### `.golangci.reference.yml`

- [ ] The linter must be added to the list of available linters (alphabetical case-insensitive order).
    - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
    - The values must be different from the default ones.
    - The default values must be defined in a comment.
    - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `Manager.GetAllSupportedLinterConfigs(...)` and `.golangci.reference.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
    - if the linter doesn't use types: `goanalysis.LoadModeSyntax`
    - `goanalysis.LoadModeTypesInfo` required `WithLoadForGoAnalysis()` in the `Manager.GetAllSupportedLinterConfigs(...)`
- [ ] The version in `WithSince(...)` must be the next minor version (`v1.X.0`) of golangci-lint.
- [ ] `WithURL()` must contain the URL of the repository.

### Recommendations

- [ ] The linter should not use SSA. (SSA can be a problem with generics)
- [ ] The linter repository should have a readme and linting.
- [ ] The linter should be published as a binary. (useful to diagnose bug origins)

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

The code review will start after the completion of those checkboxes (except for the specific items that the team will help to verify).

If the author of the PR is a member of the golangci-lint team, he should not edit this message.

**This checklist does not imply that we will accept the linter.**

**Comment by Antonboom** [comment]
Related to

https://github.com/golangci/golangci-lint/issues/3722

**Comment by jjti** [comment]
> Related to
> 
> #3722

Hey, @Antonboom. I totally agree they're related. Both check usage of [OpenTelemetry traces](https://pkg.go.dev/go.opentelemetry.io/otel/trace) (which are difficult to use and get right). They have quite different goals and checks, though. I tried to describe the differences below.

#### Same

- relate to [OpenTelemetry traces](https://pkg.go.dev/go.opentelemetry.io/otel/trace)

#### Different

The summary of below is: `go-spancheck` checks that spans are used "correctly" after they're created. `End()` is called and, for the purists that enable it, `SetStatus` and `RecordError` are called as well. `go-opentelemetry-lint` checks that they're created in the first place.

- instantiation: `polyfloyd/go-opentelemetry-lint` throws warnings for every function that does not create a span. `go-spancheck` does not
- usage - `End()`: `go-spancheck` checks that `span`s are `End()`'ed, which helps prevent memory leaks. I've seem memory leaks of this flavor crash services from OOM-kills. Calling `End` is important and called out as such in the OTEL SDK
- usage - `SetStatus()`: `go-spancheck` checks that `SetStatus` is called when a function with a span returns an error. [The rational for this is described in the README](https://github.com/jjti/go-spancheck?tab=readme-ov-file#spansetstatuscodeserror-msg-check). In short: retention and searching for `error` traces is often based on `status`
- usage - `RecordError()`: `go-spancheck` checks that `RecordError()` is called with a function with a span returns an error.  [This is described in the README](https://github.com/jjti/go-spancheck?tab=readme-ov-file#spanrecorderrorerr-check). This creates an event on the span that also helps with debugging. 

##### Different goals

As an aside, I do not agree with the goal of `go-opentelemetry-lint`. Not every function that takes a `context.Context` should have a span. That leads to large traces, with lots of spans, and lots of OTel span boilerplate, but without adding any context besides reproducing the call stack in traces -- which I would argue is not useful since developers can trace their call stack by looking at their code.

My personal and subjective experience has been that a small number of spans, each liberally annotated with `SetAttributes`, `SetStatus`, and `RecordError`, is more useful than many sparsely annotated spans. Creating a span in every single function that takes a `context.Context` is verbose -- lots of lines of code -- without a corresponding benefit (debugging production issues and investigating requests doesn't become 6x easier when a trace has 30 spans vs 5. If anything it becomes _more_ difficult because more spans have to be clicked on and investigated to find the attributes that matter for a given request.

**Comment by ldez** [comment]
The names of the options (`enable-all`, `disable-end-check`, `enable-record-error-check`, `enable-set-status-check`) are a bit weird:
- there is no list of checks to enable, and all checks (except one) are enabled by default, so `enable-all` seems off-topic and gives the feeling of unneeded complexity. And the fact this option overrides settings is confusing.
- the option names that start with `enable-` or `disable-` are a bit redundant, I expect something like `end-check`, `record-error-check`, `set-status-check` with a simple boolean with the right default (`true` or `false`).

**Comment by jjti** [comment]
> * there is no list of checks to enable

Point taken re: there being no list, and the settings being a bit verbose. I just changed settings to leave have only a `checks` list of strings. Now the settings are down to just two fields:

```go
type SpancheckSettings struct {
	Checks                []string `mapstructure:"checks"`
	IgnoreCheckSignatures []string `mapstructure:"ignore-check-signatures"`
}
```

**Comment by jjti** [comment]
@ldez [I see the `slices` issue with go 1.20](https://github.com/golangci/golangci-lint/actions/runs/7392953453/job/20112172220?pr=4290), fixing now

**Comment by ldez** [comment]
You are using the package `slices` that is not available in go1.20 (the current min version for golangci-lint)
https://github.com/jjti/go-spancheck/blob/e6faa1d972f28050ae0dc13c79ab6f75021e5df2/config.go#L8

Also, your CI configuration should use a fixed minor version of Go to ensure the Go min version of your linter, the current CI configuration doesn't allow that:
https://github.com/jjti/go-spancheck/blob/e6faa1d972f28050ae0dc13c79ab6f75021e5df2/.github/workflows/ci.yaml#L11

**Comment by ldez** [comment]
I recommend using `1.20` (fixed minor version) or `oldstable` instead of a fixed "patched" version.
https://github.com/jjti/go-spancheck/blob/40b71645c36eea47962a73e061ef261fbd2d5a4d/.github/workflows/ci.yaml#L11C15-L11C22

**Comment by jjti** [comment]
> I recommend using `1.20` (fixed minor version without patch) or `oldstable`

Makes sense and thanks for that tip, I was unaware of that alias. Just made the switch in ci and [confirmed it passed for spancheck v0.4.2](https://github.com/jjti/go-spancheck/actions/runs/7393068422) which doesn't have the `slices` pkg

**Comment by ldez** [comment]
It's always frustrating not to be able to use the latest language features, but this is the price of the compatibility with the previous Go version.
when go1.22 will be GA, I hope soon, you (and me :smile_cat:) will be able to use the `slices` package.

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**